### PR TITLE
Support for reading various media file formats beside `mraw`

### PR DESCRIPTION
**Pull request by itomac**
The `pyidi` is upgraded with the class `VideoReader` which handles reading of frames from the video recording or file stream. It is based on the `imageio.v3` python library which supports various file formats such as:
```
SUPPORTED_IMAGE_FORMATS = ['png', 'tif', 'tiff', 'bmp', 'jpg', 'jpeg']
PYAV_SUPPORTED_VIDEO_FORMATS = ['avi', 'mkv', 'mp4', 'mov', 'm4v', 'wmv', 'webm', 'flv', 'ogg', 'ogv']
```
The access to video data is established with `cih(x)` file, where metadata about video recording are stored.
The method `_get_cih()` is copied form the `pyMRAW`. Copying was necessary because in `pyMRAW` this function also performs checking for the specified file formats and rejects execution if the condition is not met. Here, only, that part is commented out and checking is performed in `__init__` method.

The use of `mraw` files remained the same with respect to `napari` gui, with exception that works only for `mraw` files, for now.

### CODE REVIEW

**Comment by itomac** [comment]
The `VideoReader` module imported in `pyIDI` as a separate object. Added tests for identification of displacements form stream of images.

**Comment by jankoslavic** [comment]
Thank you @itomac ! I will ask @domengorjup and @klemengit for a review of this PR.

**Comment by klemengit** [comment]
@itomac, I was talking to @jankoslavic and we were thinking; do you really need to define the `cih` file when, for example, processing multiple PNG images (like in the test example)?

I guess most of the parameters from the `cih` header file could be determined based on the images themselves. Maybe the only unknown would be the sample rate, but that is not required for the displacement identification. Additionally, the `cih` file is Photron's header file, so I don't know if it makes sense to use it for general purpose. I think that TOML files would be a better fit for this, but I don't think that the header files are needed at all.

What do you think? Would you be willing to implement the proposed change?

File: pyidi/methods/_simplified_optical_flow.py [hunk 1]
```diff
@@ -172,15 +172,20 @@ def pixel_shift_fun(self, i, points, image_shape):
...
         """
-        reference_image = np.mean([self.subset(image_, subset_size)
-                                   for image_ in images], axis=0)
+        if self.pixel_shift:
+            reference_image = self.subset(self.video.reader.get_frame(self.reference_range[0]), self.subset_size)
+        else:
+            reference_image = np.zeros((self.video.reader.image_height, self.video.reader.image_width), dtype=float)
+            for frame in range(self.reference_range[0], self.reference_range[1]):
+                reference_image += self.subset(self.video.reader.get_frame(frame), self.subset_size)
+            reference_image /= (self.reference_range[1] - self.reference_range[0])
```
Comment (ThijsIllimited) [inline pyidi/methods/_simplified_optical_flow.py:189]: @itomac I can't make any changes, but something like this should do it:
```
for frame in range(self.reference_range[0], self.reference_range[1]):
    reference_image += self.video.reader.get_frame(frame)
reference_image /= (self.reference_range[1] - self.reference_range[0])
reference_image = self.subset(reference_image, self.subset_size)
```

Comment (itomac) [reply pyidi/methods/_simplified_optical_flow.py:189]: Great, thanks for update.

**Comment by jankoslavic** [comment]
@itomac can we proceed with this PR towards integrating it?

**Comment by itomac** [comment]
I have resolved the conflict, but I have not pulled it yet. I am working on implementation of metadata via `.toml` file.
This is the example:
```toml
# Example of .toml file for PyIDI

title = "Test sequence for PyIDI"

[owner]
name = "Janko Slavič (janko.slavic@fs.uni-lj.si)"
dob = 2016-08-31T12:00:00+00:00

[sensor]
camera_Type = "Blender 3.3"
description = "Sequence of images to test videoreader method in PyIDI"
file_format = "PNG"
start_frame = 0
total_frame = 10

[sensor.color]
color_type = "Mono"
color_bit = 16
use_channel = "Y" # color channel (R, G, B, Y) to be used for the analysis in case of color images 
```

I propose to switch pyidi to `.toml` and format used is defined with `file_format = ` option, e.g. `"CIX"`, `"PNG"`, `"MP4"`, ...
Some options could be also set with this file, e.g. if images have three channels, then with option `use_hannel` selection can be predefined. What do you think?

**Comment by jankoslavic** [comment]
having toml file for additional parameters sounds very good to me. @klemengit ?

**Comment by klemengit** [comment]
Yes, I think this is a good idea. 
In some many cases, though, the config file would probably not be needed? So if someone just has e.g. 20 images in a folder, it would make sense to just add the path to the folder and the "video" would be created?
Off course, if one would like to add additional information, the `toml` is perfect.

**Comment by itomac** [comment]
I have made the upgrade and made pyIDI accept other file formats then Phortron camera header file. Implementation is made according to @klemengit suggestion, without extra metafile.
It is enough to provide a path to any file in stream, and reader detect that it is a single image and then automatically lists all files in the directory with the same extension, e.g. `PNG`. Images in the stream has to be named apropriately to appear sorted in the list, e.g. for image stream of 10000 images, files should be named: `image_0000.png` - `image_9999.png`
Files with multiple images such as, `tif`, `gif` are also supported.

The conflict from @ThijsIllimited push is resolved.

**Comment by klemengit** [comment]
Thanks @itomac !

I added a test for `np.ndarray` and also modified the code so that it now doesn't return an error when passed a numpy array. I ran the tests and everything is passing now.
I also updated the documentation.

There are, however, two conflicts with the current master branch. The conflicts are:
- `pyidi/methods/_simplified_optical_flow.py` in the `_reference` method,
- `requirements.txt`

Can @ThijsIllimited  and @itomac  take a look at that conflict? Especially the `_reference` method, to make sure that both functionalities are preserved. Thanks!

**Comment by itomac** [comment]
Thanks @klemengit for the code upgrade. 

1. conflict is resolved, current code can be substituted with one provided here. Let @ThijsIllimited make final check.
2. for this conflict, I have upgraded the `requirements.txt` with `imageio[pyav]` and it can be merged with one in the master. One comment: In the master there is a `imageio[ffmpeg]` used by @kristofcufar, which makes conflict. I think that this can be also changed to `pyav`: check this [https://imageio.readthedocs.io/en/stable/_autosummary/imageio.plugins.pyav.html](imageio.plugins.pyav).

**Comment by jankoslavic** [comment]
Thank you @itomac . @kristofcufar can you please try to help?

**Comment by ThijsIllimited** [comment]
I agree with the way the reference_image is calculated. I do have an edit though

Note that _reference() is called with reference_range=(0, 100) when setting the method. When trying to do this with a video shorter than 100 frames an exception gets raised.

Also, I think we should not decide for the user not to average the reference image when using pixel shifting. You can still average the reference_image for frames where you know your structure is not vibrating yet (before impact). Or we should at least warn them that the reference_image is not what they expect.

I can't push my edits, as I don't have write access, but here is my suggestion:

`
  
    if self.pixel_shift:
        reference_image = self.subset(self.video.reader.get_frame(self.reference_range[0]), self.subset_size)
        warnings.warn('Pixel shift is set to True. The reference image is set to the first frame.')
    else:
        if self.reference_range[1] >= self.video.reader.N:
            self.reference_range = (0, self.video.reader.N)
            warnings.warn('Reference range is larger than the number of images. The reference range is set to (0, N).')
        reference_image = np.zeros((self.video.reader.image_height, \
                                    self.video.reader.image_width), dtype=float)
        for frame in range(self.reference_range[0], self.reference_range[1]):
            reference_image += self.video.reader.get_frame(frame)
        reference_image /= (self.reference_range[1] - self.reference_range[0])
        reference_image = self.subset(reference_image, self.subset_size)
`

**Comment by jankoslavic** [comment]
Hi. I have added @ThijsIllimited , @itomac and @kristofcufar with write access. @ThijsIllimited can you now push it?

**Review by klemengit — APPROVED** [review]
I pushed the changes that @ThijsIllimited suggested. 
I think that this PR is ready for merge now :)

Thanks all!

**Comment by jankoslavic** [comment]
Great. Thank you Klemen. Any final thoughts by anybody involved?

**Comment by itomac** [comment]
Thank you @ThijsIllimited, I absolutely agree with your comment. I have removed checking if pixel shifting is enabled. Now, user has full control of the reference range.
Also, it is a great that you have added this checking of the reference range right limit with total number of pixels!

**Comment by klemengit** [comment]
Ok, I now resolved the conflict, `imageio[pyav]` is now used (@kristofcufar confirmed that it's ok). The `_reference` is now what @itomac last pushed.

@ThijsIllimited do you have any final comments?

**Review by ThijsIllimited — APPROVED** [review]

**Comment by jankoslavic** [comment]
Ok. Thank you everybody for making this PR reality!

**Pull request merged**